packages/xstate-react/src/useMachine.ts

```diff
@@ -81,21 +76,36 @@ export function useMachine<
         nextState.changed === undefined &&
         Object.keys(nextState.children).length;
 
-      if (nextState.changed || initialStateChanged) {
-        setState(nextState);
-      }
+      return !(nextState.changed || initialStateChanged);
+    },
+    [service]
+  );
+
+  const subscribe = useCallback(
+    (handleStoreChange) => {
+      const { unsubscribe } = service.subscribe(handleStoreChange);
+      return unsubscribe;
     },
-    []
+    [service]
+  );
+  const storeSnapshot = useSyncExternalStoreWithSelector(
+    subscribe,
+    getSnapshot,
+    undefined,
```

Comment:
This is an argument slot for `getServerSnapshot`. I'm not yet sure how this should work... maybe we should just use the initialState for it? It somewhat doesn't matter for `useMachine` because nothing should be able to change its state between render and the commit phase but all the other hooks are also using the `useSyncExternalStore` stuff and the same cannot be said about them - so we should think how this should be handled there.